examples/sparse_regression.py

```diff
@@ -0,0 +1,320 @@
+
+Note that in contrast to [1] we do MAP estimation for the kernel hyperparameters
+instead of HMC. This is not expected to be as robust as doing full Bayesian inference,
+but in some regimes this works surprisingly well. For the latter HMC approach see
+the NumPyro version:
+
+https://github.com/pyro-ppl/numpyro/blob/master/examples/sparse_regression.py
+
+References
+[1] The Kernel Interaction Trick: Fast Bayesian Discovery of Pairwise
+    Interactions in High Dimensions.
+    Raj Agrawal, Jonathan H. Huggins, Brian Trippe, Tamara Broderick
+    https://arxiv.org/abs/1905.06501
+"""
+
+
+pyro.enable_validation(True)
+torch.set_default_tensor_type('torch.FloatTensor')
+
+
+def dot(X, Z):
+    return torch.mm(X, Z.t())
+
+
+# The kernel that corresponds to our quadratic regressor.
+def kernel(X, Z, eta1, eta2, c):
+    eta1sq, eta2sq = eta1.pow(2.0), eta2.pow(2.0)
+    k1 = 0.5 * eta2sq * (1.0 + dot(X, Z)).pow(2.0)
+    k2 = -0.5 * eta2sq * dot(X.pow(2.0), Z.pow(2.0))
+    k3 = (eta1sq - eta2sq) * dot(X, Z)
+    k4 = c ** 2 - 0.5 * eta2sq
+    return k1 + k2 + k3 + k4
+
+
+# Most of the model code is concerned with constructing the sparsity inducing prior.
+def model(X, Y, hypers, jitter=1.0e-4):
+    S, P, N = hypers['expected_sparsity'], X.size(1), X.size(0)
+
+    sigma = pyro.sample("sigma", dist.HalfNormal(hypers['alpha3']))
+    phi = sigma * (S / math.sqrt(N)) / (P - S)
+    eta1 = pyro.sample("eta1", dist.HalfCauchy(phi))
+
+    msq = pyro.sample("msq", dist.InverseGamma(hypers['alpha1'], hypers['beta1']))
+    xisq = pyro.sample("xisq", dist.InverseGamma(hypers['alpha2'], hypers['beta2']))
+
+    eta2 = eta1.pow(2.0) * xisq.sqrt() / msq
+
+    lam = pyro.sample("lambda", dist.HalfCauchy(torch.ones(P, device=X.device)).to_event(1))
+    kappa = msq.sqrt() * lam / (msq + (eta1 * lam).pow(2.0)).sqrt()
+    kX = kappa * X
+
+    # sample the observation noise
+    var_obs = pyro.sample("var_obs", dist.InverseGamma(hypers['alpha_obs'], hypers['beta_obs']))
+
+    # compute the kernel for the given hyperparameters
+    k = kernel(kX, kX, eta1, eta2, hypers['c']) + (var_obs + jitter) * torch.eye(N, device=X.device)
+
+    # observe the outputs Y
+    pyro.sample("Y", dist.MultivariateNormal(torch.zeros(N, device=X.device), covariance_matrix=k),
+                obs=Y)
+
+
+"""
+Here we compute the mean and variance of coefficients theta_i (where i = dimension) as well
+as for quadratic coefficients theta_ij for a given (in our case MAP) estimate of the kernel
+hyperparameters (eta1, xisq, ...).
+Compare to theorem 5.1 in reference [1].
+"""
+@torch.no_grad()
+def compute_posterior_stats(X, Y, msq, lam, eta1, xisq, c, var_obs, jitter=1.0e-4):
+    N, P = X.shape
+
+    # prepare for computation of posterior statistics for singleton weights
+    probe = torch.zeros((2, P, P), dtype=X.dtype, device=X.device)
+    probe[0, ...] = torch.eye(P, dtype=X.dtype, device=X.device)
+    probe[1, ...] = -torch.eye(P, dtype=X.dtype, device=X.device)
```

Comment:
+1 another cool observation! It would be nicer if we can make `probe` have shape `(P, 2, P)` to match the pairwise part (it is also easier to think that we are create P probes of shapes `(2, P)`).

Reply:
done